graylog2-server/src/main/java/org/graylog2/cluster/lock/MongoLockService.java

```diff
@@ -89,6 +93,18 @@ public Optional<Lock> lock(@Nonnull String resource) {
         return lock(resource, null);
     }
 
+    @Override
+    public Optional<Lock> lock(@Nonnull String resource, int maxConcurrency) {
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(resource));
+        final long l = collection.countDocuments(eq(FIELD_RESOURCE, resource));
```

Comment:
Performing a query and then attempting to lock based on the result is prone to race conditions. We can't do that.

An approach that should work has been outlined [here](https://github.com/Graylog2/graylog-plugin-enterprise/issues/6313#issuecomment-1878646487).

Reply:
Trying to be too clever and limiting the number of DB queries. Assuming that concurrency limits are not that common, and that the backoff duration between retries is reasonable, we can probably get away with the sequential polling approach.